packages/contracts/contracts/L1/ValidatorPool.sol

```diff
@@ -105,72 +93,196 @@ contract ValidatorPool is ReentrancyGuardUpgradeable, Semver {
+            bonding.amount = 0;
+
+            return true;
         }
 
-        emit Deposit(msg.sender, msg.value, newDeposits);
+        return false;
     }
 
-    function withdraw(uint256 amount) external nonReentrant {
-        uint256 deposits = depositsOf[msg.sender];
-        require(deposits >= amount, "ValidatorPool: insufficient deposits");
+    function _increaseBalance(address _validator, uint256 amount) private {
+        uint256 balance = balances[_validator] + amount;
 
-        (bool success, ) = address(this).call{ value: amount }("");
-        require(success, "ValidatorPool: failed to withdraw");
+        if (balance >= MIN_BONDING_AMOUNT) {
+            if (validators.length == 0 || validators[0] != _validator) {
+                validatorIndexes[_validator] = validators.length;
+                validators.push(_validator);
+            }
+        }
 
-        emit Withdraw(msg.sender, amount, deposits - amount);
+        balances[_validator] = balance;
     }
 
-    function slash(address validator, address challenger) external nonReentrant {
-        require(
-            msg.sender == COLOSSEUM_ADDRESS,
-            "ValidatorPool: sender is not a colosseum contract"
-        );
-        require(
-            validator != TRUSTED_VALIDATOR,
-            "ValidatorPool: cannot slash deposits of trusted validator"
-        );
+    function _decreaseBalance(address _validator, uint256 _amount) private {
+        uint256 balance = balances[_validator];
+        require(balance >= _amount, "ValidatorPool: insufficient balances");
+        balance = balance - _amount;
 
-        uint256 validatorDeposits = depositsOf[validator];
-        uint256 slashing = SLASHING_AMOUNT;
+        if (balance < MIN_BONDING_AMOUNT) {
+            uint256 lastValidatorIndex = validators.length - 1;
+            if (lastValidatorIndex > 0) {
+                uint256 validatorIndex = validatorIndexes[_validator];
+                address lastValidator = validators[lastValidatorIndex];
 
-        if (validatorDeposits < SLASHING_AMOUNT) {
-            slashing = validatorDeposits - SLASHING_AMOUNT;
+                validators[validatorIndex] = lastValidator;
+                validatorIndexes[lastValidator] = validatorIndex;
+            }
+            delete validatorIndexes[_validator];
+            validators.pop();
         }
 
-        depositsOf[validator] -= slashing;
-        depositsOf[challenger] += slashing;
+        balances[_validator] = balance;
+    }
 
-        emit Slashing(validator, challenger, validatorDeposits);
+    /**
+     * @notice Returns output submitter address.
+     *
+     * @param _outputIndex Index of L2 checkpoint output.
+     *
+     * @return Address of output submitter.
+     */
+    function getOutputSubmitter(uint256 _outputIndex) internal view returns (address) {
+        Types.CheckpointOutput memory output = L2_ORACLE.getL2Output(_outputIndex);
+        return output.submitter;
     }
```

Comment:
It's implemented here because there's nowhere else that requires an output submitter.

Reply:
I thought that it would be more appropriate that the functions getting information from `L2OutputOracle` exist within `L2OutputOracle`, then import and use the functions where they are needed.